### ISSUE TITLE
Error Mason2 parsing VCF file v4.2 with multiple formats in the `FORMAT` field

### ISSUE THREAD

**Issue by maxrossi91** [seqan/seqan#2475]
I was trying to use variations from the GIAB dataset to simulate the reads, however, I got the following error.

```
MASON SIMULATOR
===============

____INITIALIZING______________________________________________________________

Opening reference and variants file ... OK
Initializing fragment-to-contig distribution ... OK
Initializing simulation threads ... OK
Opening output file reads/Chr21.1M.HG002.R1.fastq ... OK
Opening output file reads/Chr21.1M.HG002.R2.fastq ... OK
Opening output file sams/Chr21.1M.HG002.mates.sam... OK

____READ SIMULATION___________________________________________________________

Distributing fragments to 1 contigs (2 haplotypes each) ... OK

Simulating Reads:
/home/osboxes/seqan/include/seqan/basic/basic_exception.h:363 FAILED!  (Uncaught exception of type seqan::BadLexicalCast: Unable to convert '1:.:31:11,7:9,5:177' into unsigned int.)

stack trace:
  0                      [0x41022c]
  1                      [0x566120]
  2                      [0x56d4e6]
  3                      [0x56d531]
  4                      [0x578a78]
  5                      [0x4eb735]
  6                      [0x55976d]
  7                      [0x558936]
  8                      [0x5584df]
  9                      [0x55809d]
 10                      [0x429b2c]
 11                      [0x42a115]
 12                      [0x429150]
 13                      [0x40e728]
 14                      [0x5e063b]
 15                      [0x402901]
```
It seems to me that Mason2 is not able to correctly parse the `FORMAT` column and the `HG002` column correctly, e.g., (`INFO` ellipsed)

```
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  HG002
chr21   9602367 .       GTGTC   G       50      PASS    pla...  GT:PS:DP:ADALL:AD:GQ    0/1:.:31:11,7:9,5:177
```

If necessary I can provide the list of commands to reproduce the error. 

Mason installed via conda using `conda install -c bioconda mason -y` with version:
```
mason_simulator version: 2.0.9 [e165baf]
SeqAn version: 2.4.0
```

Thank you very much for your help.

**Comment by smehringer** [comment]
Hi @maxrossi91,

happy to hear you are using one of our Apps.

The error seems to be somewhere else. I tried to reproduce your error with `ref.fa`:
```
> ref
ACGACTAGGCTAGCTAGTCGATCGTCGATCGTAGCTAGCTAGCTGATCGATCVTAGCATGCTACGATGTACTGTGCATG
ACGACTAGGCTAGCTAGTCGATCGTCGATCGTAGCTAGCTAGCTGATCGATCVTAGCATGCTACGATGTACTGTGCATG
ACGACTAGGCTAGCTAGTCGATCGTCGATCGTAGCTAGCTAGCTGATCGATCVTAGCATGCTACGATGTACTGTGCATG
ACGACTAGGCTAGCTAGTCGATCGTCGATCGTAGCTAGCTAGCTGATCGATCVTAGCATGCTACGATGTACTGTGCATG
ACGACTAGGCTAGCTAGTCGATCGTCGATCGTAGCTAGCTAGCTGATCGATCVTAGCATGCTACGATGTACTGTGCATG
ACGACTAGGCTAGCTAGTCGATCGTCGATCGTAGCTAGCTAGCTGATCGATCVTAGCATGCTACGATGTACTGTGCATG
```
And the offending variant in the `ref.vcf`:
```
##fileformat=VCFv4.1
##fileDate=20090805
##source=myImputationProgramV3.1
##reference=ref.fa
##contig=<ID=ref,length=481,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species="Homo sapiens",taxonomy=x>
##phasing=partial
##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
##FORMAT=<ID=ADALL,Number=1,Type=Integer,Description="Haplotype Quality">
##FORMAT=<ID=AD,Number=1,Type=Integer,Description="Haplotype Quality">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Haplotype Quality">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	HG002
ref	7	.	A	G	50	PASS	NS=3	GT:PS:DP:ADALL:AD:GQ	0/1:.:31:11,7:9,5:177
```

But the program runs fine. 

```
foo@foo$ ./bin/mason_simulator -ir ref.fa -iv ref.vcf -n 1 -o out.fa
MASON SIMULATOR
===============

____INITIALIZING______________________________________________________________

Opening reference and variants file ... OK
Initializing fragment-to-contig distribution ... OK
Initializing simulation threads ... OK
Opening output file out.fa ... OK

____READ SIMULATION___________________________________________________________

Distributing fragments to 1 contigs (2 haplotypes each) ... OK

Simulating Reads:
   (allele 1) . (0 fragments) OK
   (allele 2) .. (1 fragments) OK
  Done simulating reads.

Joining temporary files ... OK
```

Could you provide me with a minimal example I can check the error with?

**Comment by eseiler** [comment]
Hey @maxrossi91,

thank you for the example, it triggers the bug for me.
The bug happens on both `master` (the latest release) and `develop` (all commits since latest release).

I'll investigate the bug and provide a patch.

Would it be possible for you to build `mason_simulator` locally once I patched the bug?
Alternatively, I could provide a precompiled binary until we do the next release and update the conda packages.

**Comment by maxrossi91** [comment]
Hi @eseiler,

Thanks for the prompt reply. I forgot to mention that I already have a workaround for this problem, which is filtering the VCF file using the following python script.

### Fix the `FORMAT` and `HG002` column
Mason2 is not able to properly parse GIAB's VCF so we need to remove some information in the `FORMAT` column and in the `HG002` column using the following script.
```python
filename = 'vcfs/HG002_21.vcf'

with open(filename, 'r') as fp:
   for l in fp:
        if l[0] == '#':
            print(l, end="")
        else:
            s = l.split('\t')
            s[8] = (s[8].split(':'))[0]
            s[9] = (s[9].split(':'))[0]
            print('\t'.join(s))
```

```bash
python3 vcf_filter.py > vcfs/HG002_21.filt.vcf
```

So I can move forward with my experiments, but I thought it might be helpful to report the problem anyway :)

**Comment by eseiler** [comment]
> So I can move forward with my experiments, but I thought it might be helpful to report the problem anyway :)

Yes! It's definitely helpful :)
Thank you for the workaround, it should help narrow the problem further down.

**Comment by eseiler** [comment]
This'll be my post for observations:

When running
```bash
bcftools view HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz -Ov --regions chr21 > HG002_21.vcf
perl -n -i.bak -e 'print unless m/##contig=<ID=chr(?!21)/'  HG002_21.vcf
```
All contigs but `chr21` are removed from the vcf.

However, mason gets the number of contigs from the fasta index:
https://github.com/seqan/seqan/blob/3d34785d4aa3ec7e032a5c72ff9e9dac2511e5f5/apps/mason2/mason_simulator.cpp#L1095

There are 195 contigs in the fai, but only 1 in the vcf.

Modifying it to:
```cpp
        size_t const debug_fai = numSeqs(vcfMat.faiIndex);
        size_t const debug_vcf = length(contigNames(context(vcfMat.vcfFileIn)));
        for (unsigned i = 0; i < std::min(debug_fai, debug_vcf); ++i)
```
results in a `lexicalCast` fail as described in first post:
```
Uncaught exception of type seqan::BadLexicalCast: Unable to convert '1:.:31:11,7:9,5:177' into unsigned int.
```

**Comment by eseiler** [comment]
There seem to be 2 problems:

* VCF and FASTA must contain the same number of records, but this is never checked.
* For indels, the parsing of `1/1:.:31:11,7:9,5:177` does not stop after the first colon, like it does for all the SNP cases.

**Comment by eseiler** [comment]
This should be fixed on `develop` by #2476 

There were two problems:

* Mason expects the FASTA reference and VCF file to contain equally many contigs
-> Now throws an error with some explanation. The optimal solution would be to apply the variants selectively to the given contigs, and use the rest as is. However, this is a more involved fix, which we do not have the resources for.

* In some cases, the parsing of the FORMAT column did not stop after the genotype
-> Should be fixed. Your example should now work without preprocessing from your side.
